### PR TITLE
lib-prep recalibration pipelines

### PR DESCRIPTION
**Pull request by doron-st**
This code is still research, but since @kenissur me and @rinashalibo are working on the same code, we must sync with git.
Pushing this to bioinfo is problematic because of the imports from ugvc.

### CODE REVIEW

File: ugvc/pipelines/lpr/filter_vcf_with_lib_prep_recalibration_model.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    )
+
+    sp.print_and_run_clt(
+        featuremap_to_dataframe.run,
+        [],
+        {"-i": featuremap_on_calls_vcf, "-r": args.ref_fasta, "-f": "TGCA", "-o": featuremap_on_calls_parquet},
+    )
+
+    tp_positions = f"{out_dir}/true_positions.csv"
+    sp.print_and_run(f"bcftools view {truth_vcf} -H -f PASS | " "awk '$6 > 10 {print $2}' > " + tp_positions)
```
Comment (kenissur) [inline ugvc/pipelines/lpr/filter_vcf_with_lib_prep_recalibration_model.py]: You can add "-i 'QUAL>10'" to bcftools instead of using awk, seems healthier to me.

Also, are you excluding 10 on purpose?

And maybe maybe it a parameter instead of hard coded?

File: ugvc/pipelines/lpr/filter_vcf_with_lib_prep_recalibration_model.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        f"-p work_dir {out_dir} "
+        f"-p vcf_input {calls_parquet} "
+        f"-p top_n_read_scores_input "
+        f"{out_dir}/recalibration_outputs/{model_name}.scored_alleles.parquet "
+        f"-p tp_positions_file {tp_positions} "
+    )
+
+    sp.print_and_run(
+        f"jupyter nbconvert --to html {out_dir}/score_vcf_with_lib_prep_calibration_data_{out_suffix}.nbconvert.ipynb"
+        f" --template full --no-input --output score_vcf_with_lib_prep_calibration_data_{out_suffix}.html"
```
Comment (kenissur) [inline ugvc/pipelines/lpr/filter_vcf_with_lib_prep_recalibration_model.py]: You can drop the "--template full", it's a deprecated syntax

File: ugvc/pipelines/lpr/train_lib_prep_recalibration_model.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    )
+
+
+def localize_inputs(args, sp):
+    if args.substitution_analysis_wid:
+        featuremap_vcf = (
+            f"gs://cromwell-backend-ultima-data-307918/cromwell-execution/SubstitutionAnalysis/"
+            f"{args.substitution_analysis_wid}/call-FeatureMap/FeatureMap/*/"
+            f"call-FeatureMapMerge/*.featuremap.vcf.gz"
+        )
```
Comment (kenissur) [inline ugvc/pipelines/lpr/train_lib_prep_recalibration_model.py]: I usually do something like this:
`
import json

with open(cloud_sync(f"gs://cromwell-backend-ultima-data-307918/cromwell-execution/SubstitutionAnalysis/{args.substitution_analysis_wid}/metadata.json")) as fh:
    metadata = json.load(fh)
featuremap_vcf = metadata["outputs"]["featuremap"]
`

File: ugvc/pipelines/lpr/train_lib_prep_recalibration_model.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        featuremap_on_calls = f"{out_dir}/featuremap_on_calls.vcf.gz"
+        sp.print_and_run(f"bcftools view --type snps -f PASS {unfiltered_calls} -Oz -o {calls_vcf}")
+        sp.print_and_run(
+            f"bedtools intersect -a {featuremap_vcf} -b {calls_vcf} -header | "
+            f"bcftools view -Oz -o {featuremap_on_calls}"
+        )
+        featuremap_vcf = featuremap_on_calls
+        sp.print_and_run(
+            f"bcftools query -f '%CHROM %POS [%VAF]\\n' {calls_vcf} | grep -v , | "
+            + "awk -v OFS='\t' '{print $1,$2-1,$2,$3}' "
```
Comment (kenissur) [inline ugvc/pipelines/lpr/train_lib_prep_recalibration_model.py]:
```suggestion
            bcftools query -f '%CHROM\t%POS0\t%POS\t[%VAF]\n' {calls_vcf}
```
Would do the trick without the awk command

File: ugvc/pipelines/lpr/train_lib_prep_recalibration_model.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    tp_args = [f"{out_dir}/estimated_tps.{suffix}" for suffix in ("bed", "vcf.gz", "parquet")]
+    fp_args = [f"{out_dir}/estimated_fps.{suffix}" for suffix in ("bed", "vcf.gz", "parquet")]
+    tp_args.extend([args.tp_min_af, 1])
+    fp_args.extend([0, args.fp_max_af])
+    afs_bed = f"{out_dir}/afs.bed"
+
+    if args.calls_vcf:
+        calls_vcf = f"{out_dir}/calls.snps.pass.vcf.gz"
+        featuremap_on_calls = f"{out_dir}/featuremap_on_calls.vcf.gz"
+        sp.print_and_run(f"bcftools view --type snps -f PASS {unfiltered_calls} -Oz -o {calls_vcf}")
```
Comment (kenissur) [inline ugvc/pipelines/lpr/train_lib_prep_recalibration_model.py]: Maybe add "-m2 -M2" to make sure you only retain biallelic variants

File: ugvc/pipelines/lpr/train_lib_prep_recalibration_model.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+        f"{out_dir}/train_lib_prep_xgboost{args.output_suffix}.nbconvert.ipynb "
+        f"-p work_dir {out_dir} "
+        f"-p suffix {output_suffix_arg} "
+        f"-p balance_motifs {balance_motifs} "
+        f"-p balance_tp_fp {balance_tp_fp} "
+    )
+
+    sp.print_and_run(
+        "jupyter nbconvert --to html "
+        f"{out_dir}/train_lib_prep_xgboost{args.output_suffix}.nbconvert.ipynb --template full "
```
Comment (kenissur) [inline ugvc/pipelines/lpr/train_lib_prep_recalibration_model.py]: --template full
can be removed

File: ugvc/pipelines/lpr/filter_vcf_with_lib_prep_recalibration_model.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        {"-i": featuremap_on_calls_vcf, "-r": args.ref_fasta, "-f": "TGCA", "-o": featuremap_on_calls_parquet},
+    )
+
+    tp_positions = f"{out_dir}/true_positions.csv"
+    sp.print_and_run(f"bcftools view {truth_vcf} -H -f PASS | " "awk '$6 > 10 {print $2}' > " + tp_positions)
+
+    base_dir = args.notebooks_dir
+    model_name = os.path.basename(args.lib_prep_model_file)
+    sp.print_and_run(
+        f"papermill {base_dir}/lib_prep_recalibration_N_top_reads_per_allele.ipynb "
```
Comment (kenissur) [inline ugvc/pipelines/lpr/filter_vcf_with_lib_prep_recalibration_model.py:73]: You could move the notebooks to the repo as well

**Review by kenissur — APPROVED** [review]
Looks good. I had some suggestion that you can take or leave.

**Review by kenissur — APPROVED** [review]
The checks failing are problems in the original branch so there is no problem merging

**Comment by doron-st** [comment]
Thanks, I will improve the code according to @kenissur's comments.
Re the notebooks, since they are hard to sync anyways, I think we can wait until they are stabilized.

@ilyasoifer do you have any general objection for adding this code to the repo?

File: ugvc/pipelines/lpr/filter_vcf_with_lib_prep_recalibration_model.py (hunk 1 above)
Comment (doron-st) [reply ugvc/pipelines/lpr/filter_vcf_with_lib_prep_recalibration_model.py]: I was excluding 10 on purpose from the fresh-frozen.
Will externalize as a parameter.

File: ugvc/__main__.py [hunk 8]
```diff
@@ -12,17 +12,25 @@
     sys.path.insert(0, path)
 
 from ugvc.cnv import filter_sample_cnvs
-
+from ugvc.joint import compress_gvcf
```
Comment (ilyasoifer) [inline ugvc/__main__.py:15]: Looks unrelated

File: ugvc/pipelines/lpr/filter_vcf_with_lib_prep_recalibration_model.py (hunk 7 above)
Comment (ilyasoifer) [reply ugvc/pipelines/lpr/filter_vcf_with_lib_prep_recalibration_model.py:73]: I would prefer not to assume we have gsutil installed in the environment. We do want to get rid of it sooner or later

File: ugvc/pipelines/lpr/train_lib_prep_recalibration_model.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        f"{out_dir}/train_lib_prep_xgboost{args.output_suffix}.nbconvert.ipynb "
+        f"--no-input --output train_lib_prep_xgboost_report{args.output_suffix}.html"
+    )
+
+
+def localize_inputs(args, sp):
+    if args.substitution_analysis_wid:
+        with open(
+            cloud_sync(
+                f"gs://cromwell-backend-ultima-data-307918/cromwell-execution/"
```
Comment (ilyasoifer) [inline ugvc/pipelines/lpr/train_lib_prep_recalibration_model.py]: Hardcoded link - please remove

**Review by ilyasoifer — APPROVED** [review]
looks good, just minor comments

File: ugvc/__main__.py (hunk 8 above)
Comment (doron-st) [reply ugvc/__main__.py:15]: Yes, I rebased on master.
I suppose it's better to ommit

File: ugvc/pipelines/lpr/train_lib_prep_recalibration_model.py (hunk 9 above)
Comment (doron-st) [reply ugvc/pipelines/lpr/train_lib_prep_recalibration_model.py]: ok ok

File: ugvc/__main__.py (hunk 8 above)
Comment (doron-st) [reply ugvc/__main__.py:15]: Oh, it's the precommit's work... I can't avoid it
